### PR TITLE
rdiff-backup-delete: Use buffered subprocess gzip or buffered io

### PR DESCRIPTION
**Pull request by ikus060**
Using buffered io improve the performance when reading and writing gzip
files.

CHG: rdiff-backup-delete: improve performance on gzip files

### CODE REVIEW

**Comment by ikus060** [comment]
After more testing, zcat/gzip are still faster then python implementation. So I'm making required changes to use zcat and gzip when available and fallback to gzip with buffered io.

```
rdiff-backup original

  real	11m43.436s
  user	11m37.380s
  sys	0m4.915s
  
rdiff-backup io.BufferedReader()

  real	8m51.554s
  user	8m46.373s
  sys	0m4.263s
  
rdiff-backup subprocess zcat/gzip 

  real	3m49.299s
  user	7m42.138s
  sys	0m18.507s
```

File: src/rdiff-backup-delete [hunk 1]
```diff
@@ -25,13 +25,17 @@
 #
 
 import os
```
Comment (ericzolf) [inline src/rdiff-backup-delete:27]: Overall, the imports should be in alphabetical order.

File: src/rdiff-backup-delete [hunk 2]
```diff
@@ -25,13 +25,17 @@
...
+import re
 import sys
 
+from distutils import spawn
```
Comment (ericzolf) [inline src/rdiff-backup-delete:31]: See https://github.com/rdiff-backup/rdiff-backup/blob/master/docs/CODING.md#import

File: src/rdiff-backup-delete [hunk 3]
```diff
@@ -60,6 +64,71 @@ def _str(value):
         return value.decode('utf-8', errors='replace')
 
 
+# Check if zcat is available.
```
Comment (ericzolf) [inline src/rdiff-backup-delete:67]: Comment says zcat but executable is gzip

File: src/rdiff-backup-delete [hunk 4]
```diff
@@ -60,6 +64,71 @@ def _str(value):
...
+GZIP = spawn.find_executable('gzip')
+
+
+class _wrap_close:
```
Comment (ericzolf) [inline src/rdiff-backup-delete:71]: Class names are Camel case.

File: src/rdiff-backup-delete [hunk 5]
```diff
@@ -60,6 +64,71 @@ def _str(value):
...
+    def __getattr__(self, name):
+        return getattr(self._stream, name)
+
+    def __iter__(self):
+        return iter(self._stream)
+
+
+def _open(fn, mode):
+    """
+    Wrapp to open a file with or without compression using zcat/gzip or
```
Comment (ericzolf) [inline src/rdiff-backup-delete:105]: Wrap (one p)

Comment (ericzolf) [reply src/rdiff-backup-delete:105]: And why do you talk again about zcat?

File: src/rdiff-backup-delete [hunk 6]
```diff
@@ -60,6 +64,71 @@ def _str(value):
...
+        return buffered(open(fn, mode))
+
+    # Open file using python gzip if zcat and gzip are not available.
+    if not GZIP:
+        return buffered(gzip.open(fn, mode))
+
+    # When available, open file using subprocess gzip for better performance
+    if 'r' in mode:
+        proc = subprocess.Popen([_bytes(GZIP), b'-cd', fn],
+                                shell=False,
```
Comment (ericzolf) [inline src/rdiff-backup-delete:121]: shell=False is the default and isn't required

File: src/rdiff-backup-delete [hunk 7]
```diff
@@ -60,6 +64,71 @@ def _str(value):
...
+                                stdout=subprocess.PIPE)
+        return _wrap_close(proc.stdout, proc)
+    else:  # wb
+        proc = subprocess.Popen([GZIP],
+                                shell=False,
```
Comment (ericzolf) [inline src/rdiff-backup-delete:126]: shell=False is the default and isn't required

File: src/rdiff-backup-delete [hunk 8]
```diff
@@ -60,6 +64,71 @@ def _str(value):
         return value.decode('utf-8', errors='replace')
 
 
+# Check if zcat is available.
+GZIP = spawn.find_executable('gzip')
```
Comment (ericzolf) [inline src/rdiff-backup-delete:68]: It's an internal "constant", its name should start with underscore.

File: src/rdiff-backup-delete [hunk 9]
```diff
@@ -60,6 +64,71 @@ def _str(value):
...
+        self._proc = proc
+
+    def close(self):
+        if self._proc.stdin:
+            self._proc.stdin.close()
+        returncode = self._proc.wait()
+        if self._proc.stdout:
+            self._proc.stdout.close()
+        if returncode == 0:
+            return None
```
Comment (ericzolf) [inline src/rdiff-backup-delete:87]: Why doing this? Why not returning directly the returncode? It's mostly a moot point as I didn't see that you do anything with it, but IMHO having consistently an integer being returned is easier to handle.

**Review by ericzolf — CHANGES_REQUESTED** [review]
Only small things, looks good to me even if it hurts to need such workarounds to get the necessary performance.

**Comment by ikus060** [comment]
> Only small things, looks good to me even if it hurts to need such workarounds to get the necessary performance.

I completely agree. gzip performance in python is a shame. It one of the reason why I recommend disabling it by default and let the compression of FS do it's thing. With zstd-4 in ZFS it even boost the performance !

File: src/rdiff-backup-delete (hunk 2 above)
Comment (ikus060) [reply src/rdiff-backup-delete:31]: Be more specific here. I have no clue what is the problem. `spawn` is a module.

File: src/rdiff-backup-delete (hunk 5 above)
Comment (ikus060) [reply src/rdiff-backup-delete:105]: I first started by using zcat ang gzip, but finally figure out how to use gzip for both operation. And I didn't review my comments.

File: src/rdiff-backup-delete (hunk 4 above)
Comment (ikus060) [reply src/rdiff-backup-delete:71]: fixed

File: src/rdiff-backup-delete (hunk 6 above)
Comment (ikus060) [reply src/rdiff-backup-delete:121]: fixed

File: src/rdiff-backup-delete (hunk 8 above)
Comment (ikus060) [reply src/rdiff-backup-delete:68]: fixed

File: src/rdiff-backup-delete (hunk 2 above)
Comment (ericzolf) [reply src/rdiff-backup-delete]: Ups. Sorry, my bad. You're right.

File: src/rdiff-backup-delete (hunk 1 above)
Comment (ericzolf) [reply src/rdiff-backup-delete]: https://www.python.org/dev/peps/pep-0008/#imports doesn't say anything about it, I have a slight preference for the 2nd approach, which is also what you chose, so fine with me.

File: src/rdiff-backup-delete [hunk 10]
```diff
@@ -60,6 +63,67 @@ def _str(value):
...
+    pure-python implementation.
+    """
+    compress = fn.endswith(b'.gz')
+    buffered = io.BufferedReader if 'r' in mode else io.BufferedWriter
+    # Open file directly if compression is not required
+    if not compress:
+        return buffered(open(fn, mode))
+
+    # Open file using python gzip if zcat and gzip are not available.
+    if not GZIP:
```
Comment (ericzolf) [inline src/rdiff-backup-delete:112]: It should be _GZIP. That's apparently why the flake8 test fails.

**Review by ericzolf — APPROVED** [review]
Just fix the tests (flake8/_GZIP) and then you can merge.

**Pull request merged**